keps/sig-node/3317-user-namespaces/README.md

```diff
@@ -0,0 +1,1015 @@
+    // Note: The CRI default is POD, but the v1.PodSpec default is CONTAINER.
+    // The kubelet's runtime manager will set this to CONTAINER explicitly for v1 pods.
+    // Namespaces currently set by the kubelet: POD, CONTAINER, NODE, TARGET
+    NamespaceMode pid = 2;
+    // IPC namespace for this container/sandbox.
+    // Note: There is currently no way to set CONTAINER scoped IPC in the Kubernetes API.
+    // Namespaces currently set by the kubelet: POD, NODE
+    NamespaceMode ipc = 3;
+    // Target Container ID for NamespaceMode of TARGET. This container must have been
+    // previously created in the same pod. It is not possible to specify different targets
+    // for each namespace.
+    string target_id = 4;
+    // User namespace for this sandbox.
+	// The Kubelet picks the user namespace configuration to use for the sandbox.  The mappings
+	// are specified as part of the UserNamespace struct.  If the struct is nil, then the POD mode
+	// must be assumed.  This is done for backward compatibility with older Kubelet versions that
+	// do not set a user namespace.
+	UserNamespace userns_options = 5;
+}
+
+```
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
+either more isolation or more workloads. When no support is yet added to handle
+some workload, a clear error will be shown.
+
+PLEASE note that only phase 1 is targeted for alpha. Also note that the missing
+details (CRI changes, changes needed in container runtimes, etc.) will be added
+in follow-up PRs.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase. That table is not updated (it is from the initial
+proposal, doesn't have all the feedback and adjustments we discussed) but can
+still be useful as a general overview.
+
+TODO: move the table to markdown and include it here.
+
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volumes work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and can only be set to `false`
+on pods which use either no volumes or only volumes of the following types:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65535, mapping the range 0-65534 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65535 (2^16-1) top. This imposes a limit of 65k pods per node,
+but that is not an issue we will hit in practice for a long time, if ever (today
+we run 110 pods per node by default).
```

Comment:
How does kubelet choose the subset of host UID space that it will assign to the pod?

Do I remember correctly that user actually have a a way to choose their UID currently, so the space may in theory be fragmented?

Also - follow up question - don't we need to store the chosen subspace somewhere? If not, can you explain why (in the KEP)?